nutils/evaluable.py

```diff
@@ -1774,41 +1776,36 @@ def _intbounds_impl(self):
-        self.funcs = funcs
-        func1, func2 = funcs
-        assert equalshape(func1.shape, func2.shape) and func1.dtype == func2.dtype != bool, 'Add({}, {})'.format(func1, func2)
-        super().__init__(args=tuple(self.funcs), shape=func1.shape, dtype=func1.dtype)
+        assert isinstance(funcs, types.frozenmultiset), f'funcs={funcs!r}'
+        func, *others = funcs
+        assert all(equalshape(f.shape, func.shape) and f.dtype == func.dtype for f in others)
+        self.funcs = tuple(funcs)
+        super().__init__(args=tuple(funcs), shape=func.shape, dtype=func.dtype)
 
     @cached_property
     def _inflations(self):
-        func1, func2 = self.funcs
-        func2_inflations = dict(func2._inflations)
-        inflations = []
-        for axis, parts1 in func1._inflations:
-            if axis not in func2_inflations:
-                continue
-            parts2 = func2_inflations[axis]
-            dofmaps = set(parts1) | set(parts2)
-            if (len(parts1) < len(dofmaps) and len(parts2) < len(dofmaps)  # neither set is a subset of the other; total may be dense
-                    and self.shape[axis].isconstant and all(dofmap.isconstant for dofmap in dofmaps)):
-                mask = numpy.zeros(int(self.shape[axis]), dtype=bool)
-                for dofmap in dofmaps:
-                    mask[dofmap.eval()] = True
-                if mask.all():  # axis adds up to dense
-                    continue
-            inflations.append((axis, types.frozendict((dofmap, util.sum(parts[dofmap] for parts in (parts1, parts2) if dofmap in parts)) for dofmap in dofmaps)))
-        return tuple(inflations)
+        inflations_per_axis = [[] for i in range(self.ndim)]
+        for func in self.funcs:
+            for axis, parts in func._inflations:
+                inflations_per_axis[axis].append(parts)
+        return tuple((axis,
+            types.frozendict((dofmap, add(*(parts[dofmap] for parts in parts_per_func if dofmap in parts)))
+                for dofmap in functools.reduce(set.union, map(set, parts_per_func))))
+                    for axis, parts_per_func in enumerate(inflations_per_axis)
+                        if len(parts_per_func) == len(self.funcs))
 
     def _simplified(self):
-        func1, func2 = self.funcs
-        if func1 == func2:
-            return multiply(func1, 2)
-        for axes1 in func1._diagonals:
-            for axes2 in func2._diagonals:
-                if len(axes1 & axes2) >= 2:
-                    axes = sorted(axes1 & axes2)[:2]
-                    return diagonalize(takediag(func1, *axes) + takediag(func2, *axes), *axes)
+        for j, fj in enumerate(self.funcs):
+            if isinstance(fj, Add):
+                return add(*self.funcs[:j], *fj.funcs, *self.funcs[j+1:])
+            for i, fi in enumerate(self.funcs[:j]):
+                diags = [sorted(axesi & axesj)[:2] for axesi in fi._diagonals for axesj in fj._diagonals if len(axesi & axesj) >= 2]
```

Comment:
Swapping `Diagonalize` and `Add` for two arguments suffers from the same ordering problem as `unalign` for `Multiply`. Do we actually need this simplification (for more than two arguments) before `_optimize_for_numpy`?

Reply:
Promoting `Diagonalize` reduces flops, and as such is a relevant modification for the simplification step. It also exposes structure to follow-up operations that can propagate the structure further. The optimization step, in my view, is a more local sweep that combines and/or replaces operations for more efficient evaluation. Note also that this swap was always part of simplify; I don't believe this PR introduces any new simplifications or optimizations.